push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushyApnsSender.java

```diff
@@ -0,0 +1,245 @@
+
+
+    @Override
+    public void sendPushMessage(final Variant variant, final Collection<String> tokens, final UnifiedPushMessage pushMessage, final String pushMessageInformationId, final NotificationSenderCallback senderCallback) {
+        // no need to send empty list
+        if (tokens.isEmpty()) {
+            return;
+        }
+
+        final iOSVariant iOSVariant = (iOSVariant) variant;
+
+        final String payload;
+        {
+            try {
+                payload = createPushPayload(pushMessage.getMessage(), pushMessageInformationId);
+            } catch (IllegalArgumentException iae) {
+                logger.info(iae.getMessage(), iae);
+                senderCallback.onError("Nothing sent to APNs since the payload is too large");
+                return;
+            }
+        }
+
+        final ApnsClient apnsClient;
+        {
+            try {
+                apnsClient = buildApnsClient(iOSVariant);
+            } catch (Exception e) {
+                logger.error(e.getMessage(), e);
+                senderCallback.onError(e.getMessage());
+                return;
+            }
+        }
+
+        // connect and move on with sending in the background:
+        final Future<Void> connectFuture = connectToDestinations(iOSVariant, apnsClient);
+        connectFuture.addListener(new GenericFutureListener<Future<? super Void>>() {
+            @Override
+            public void operationComplete(Future<? super Void> future) throws Exception {
+
+                if (!future.isSuccess()) {
+
+                    final Throwable t = future.cause();
+                    logger.error("Error connecting to APNs", t);
+                    senderCallback.onError(t.getMessage());
+
+                } else if (apnsClient.isConnected()) {
+
+                    logger.debug("sending payload for all tokens to APNS");
+
+                    for (final String token : tokens) {
+
+                        final AeroGearApnsPushNotification pushNotification = new AeroGearApnsPushNotification(token, payload);
+                        final Future<PushNotificationResponse<AeroGearApnsPushNotification>> notificationSendFuture = apnsClient.sendNotification(pushNotification);
+
+                        notificationSendFuture.addListener(new GenericFutureListener<Future<? super PushNotificationResponse<AeroGearApnsPushNotification>>>() {
+                            @Override
+                            public void operationComplete(Future<? super PushNotificationResponse<AeroGearApnsPushNotification>> future) throws Exception {
+
+                                if (future.isSuccess()) {
+                                    handlePushNotificationResponsePerToken(notificationSendFuture.get());
+                                }
+                            }
+                        });
+                    }
+
+                    // we have managed to send all messages ;-)
+                    senderCallback.onSuccess();
```

Comment:
The sendNotification is asynchronous, so when we arrived here we still don't know if we successfully sent the message or not. Is this correct?